templates/lib/yarp/node.rb.erb

```diff
@@ -157,6 +189,55 @@ module YARP
     <%- end -%>
   end
 
+  # The dispatcher class fires events for nodes that are found while walking an AST to all registered listeners. It's
+  # useful for performing different types of analysis on the AST without having to repeat the same visits multiple times
+  class Dispatcher
+    # attr_reader listeners: Hash[Symbol, Array[Listener]]
+    attr_reader :listeners
+
+    def initialize
+      @listeners = {}
+    end
+
+    # Register a listener for one or more events
+    #
+    # def register: (Listener, *Symbol) -> void
+    def register(listener, *events)
+      events.each { |event| (listeners[event] ||= []) << listener }
+    end
+
+    # Walks `root` dispatching events to all registered listeners
+    #
+    # def dispatch: (Node) -> void
+    def dispatch(root)
+      queue = [root]
+
+      while (node = queue.shift)
+        case node.human
+        <%- nodes.each do |node| -%>
+        when :<%= node.human %>
+          listeners[:<%= node.human %>_enter]&.each { |listener| listener.<%= node.human %>_enter(node) }
+          queue = node.compact_child_nodes.concat(queue)
```

Comment:
Is this the correct order?
Maybe it should be `queue.concat(node.compact_child_nodes)`?

It looks like this adds child nodes before the current elements in the array (so on the "left side" i.e. near 0 index) and `queue.shift` above removes from the "left side" too.
So it looks like it behaves somewhat like a stack:
```ruby
q = [1]
q.shift

q = [2, 3].concat(q)
p q # => [2, 3]
p q.shift # => 2

q = [4].concat(q)
p q # => [4, 3]
p q.shift # => 4, the last added element is the first out in this case
# Not fully LIFO though because if it was `q = [4, 5].concat(q)` above, then 4 is out before 5
```